### PR TITLE
Zoom

### PR DESCRIPTION
**Pull request by tingxuanz**
hi @rowlandm  This is a pull request from zoom to bar_chart_with_click.
You can review the zoom function code.

### CODE REVIEW

File: examples/simple.js [hunk 1]
```diff
@@ -27,37 +30,56 @@ var clickedBarId = [];
...
+    /*
+    After load the file, all data are stored in metadata.
+    metadata is an array of objects. All bojects use same column names (eg. SampleType) as keys.
+    We want to get these keys as groupByOptions.
+    */
+    groupByOptions = Object.keys(metadata[0]);
+
+    /*We would like to remove SampleID, ReadFile1 and ReadFile2 from groupByOptions,
+      since they won't be used as groupby options.*/
+    var index1 = groupByOptions.indexOf("SampleID");
```
Comment (rowlandm) [inline examples/simple.js]: This could be an array that you loop through and remove....

File: lib/index.js [hunk 2]
```diff
@@ -152,15 +152,73 @@ module.exports = biojsvispca = function(init_options){
...
 
+setup_zoom = function(graph){
+  svg = graph.svg;
+  page_options = graph.page_options;
+  var zoom = d3.behavior.zoom()
+      .x(graph.scaleX)
+      .y(graph.scaleY)
+      .on("zoomstart", zoom_start)
+      .on("zoom", function(){
+        console.log("zoom");
```
Comment (rowlandm) [inline lib/index.js]: No console logs please!!!!

File: lib/index.js [hunk 3]
```diff
@@ -152,15 +152,73 @@ module.exports = biojsvispca = function(init_options){
...
+    .call(zoom);
+
+
+  graph.svg = svg;
+  return graph;
+};
+
+//just for debugging
+function zoom_start(){
+    console.log("zoom start");
```
Comment (rowlandm) [inline lib/index.js]: No console logs, even for debugging. Use breakpoints.

File: lib/index.js [hunk 4]
```diff
@@ -319,13 +379,21 @@ module.exports = biojsvispca = function(init_options){
...
     var svg = d3.select(options.target).append("svg")
-        .attr("width", full_width)
-        .attr("height",full_height)
+        .attr("width", scatterWidth)
+        .attr("height",scatterHeight)
+      .append("g")
+        // this is just to move the picture down to the right margin length
+        .attr("transform", "translate(" + left + "," + top + ")");
+
+    var svg2 = d3.select(options.target).append("svg")
```
Comment (rowlandm) [inline lib/index.js]: Why do we want an svg2?

File: lib/index.js [hunk 5]
```diff
@@ -335,11 +403,21 @@ module.exports = biojsvispca = function(init_options){
     graph = setup_svg(graph);
     graph = setup_x_axis(graph);
     graph = setup_y_axis(graph);
+
+    /*
+    if you append your zoom/brush after your data points, the zoom/brush overlay will catch all of the pointer events,
```
Comment (rowlandm) [inline lib/index.js:402]: Good comment!

**Review by rowlandm — CHANGES_REQUESTED** [review]

File: lib/index.js (hunk 4 above)
Comment (tingxuanz) [reply lib/index.js]: The zoom function works in the whole svg area.

If we put 2 plots in one svg element, we will find that the circles in the scatter plot can reach the bar chart area when zooming in.

So, I use one svg for scatter plot, one svg for bar chart to make sure that the zoom function only works in the scatter plot.

Comment (rowlandm) [reply lib/index.js]: Thanks Peter - but perhaps it's better to explain it in the code via comments

File: lib/index.js [hunk 6]
```diff
@@ -319,13 +368,26 @@ module.exports = biojsvispca = function(init_options){
...
-        .attr("width", full_width)
-        .attr("height",full_height)
+        .attr("width", scatterWidth)
+        .attr("height",scatterHeight)
+      .append("g")
+        // this is just to move the picture down to the right margin length
+        .attr("transform", "translate(" + left + "," + top + ")");
+
+    /*
+    The zoom function works in the whole svg area.
```
Comment (tingxuanz) [inline lib/index.js:378]: add comment about why using 2 svg elements

File: lib/index.js [hunk 7]
```diff
@@ -374,6 +374,11 @@ setup_zoom = function(graph){
         // this is just to move the picture down to the right margin length
         .attr("transform", "translate(" + left + "," + top + ")");
 
+    /*
+    The zoom function works in the whole svg area.
+    If we put 2 plots in one svg element, we will find that the circles in the scatter plot can reach the bar chart area when zooming in.
```
Comment (rowlandm) [inline lib/index.js:379]: Should have said this before. Could we then name them svg_pca and svg_bar ?

File: examples/simple.js [hunk 8]
```diff
@@ -44,17 +44,35 @@ body.appendChild(groupBySelectList); // append the select list to body
...
     groupByOptions = Object.keys(metadata[0]);
 
     //We would like to remove SampleID, ReadFile1 and ReadFile2 from groupByOptions, because they won't be used as groupby options.
-    var index1 = groupByOptions.indexOf("SampleID");
-    groupByOptions.splice(index1,1);
-    var index2 = groupByOptions.indexOf("ReadFile1");
-    groupByOptions.splice(index2,1);
-    var index3 = groupByOptions.indexOf("ReadFile2");
-    groupByOptions.splice(index3,1);
+    for (var i = 0; i < groupByOptions.length; i++) {
```
Comment (rowlandm) [inline examples/simple.js]: I was thinking more:
var array_to_remove_items = ["SampleID","ReadFile1","ReadFile2"];
etc etc

**Review by rowlandm — APPROVED** [review]

**Pull request merged**